components/json/json_64bit_support_unittest.cc

```diff
@@ -0,0 +1,104 @@
+#include "base/json/json_writer.h"
+#include "base/numerics/byte_conversions.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/values.h"
+#include "brave/components/json/schema.h"
+#include "testing/gmock/include/gmock/gmock.h"
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace brave_wallet {
+
+namespace {
+constexpr char kSampleJson[] =
+    R"a({"boolean":true,"maxint":9223372036854775807,"maxuint":18446744073709551615,"minint":-9223372036854775808,"minuint":0,"name":"whatever","price":1.5,"simpleint":2147483647})a";
+}
+
+using ::testing::Optional;
+
+TEST(Json64BitSupportTest, Write64BitValues) {
+  base::Value::Dict dict;
+  dict.Set("maxuint", base::Value(base::as_byte_span(base::NumberToString(
+                          std::numeric_limits<uint64_t>::max()))));
+  dict.Set("minuint", base::Value(base::as_byte_span(base::NumberToString(
+                          std::numeric_limits<uint64_t>::min()))));
+  dict.Set("maxint", base::Value(base::as_byte_span(base::NumberToString(
+                         std::numeric_limits<int64_t>::max()))));
+  dict.Set("minint", base::Value(base::as_byte_span(base::NumberToString(
+                         std::numeric_limits<int64_t>::min()))));
+  dict.Set("price", 1.5);
+  dict.Set("boolean", true);
+  dict.Set("name", "whatever");
+  dict.Set("simpleint", std::numeric_limits<int>::max());
+  EXPECT_THAT(
+      base::WriteJsonWithOptions(dict, base::OPTIONS_SERIALISE_64BIT_NUMBERS),
+      kSampleJson);
+}
+
+TEST(Json64BitSupportTest, Read64BitValues) {
+  auto parsed =
+      base::JSONReader::ReadDict(kSampleJson, base::JSON_ALLOW_64BIT_NUMBERS);
+  ASSERT_TRUE(parsed);
+  ASSERT_TRUE(parsed->Find("boolean"));
+  ASSERT_TRUE(parsed->Find("name"));
+  ASSERT_TRUE(parsed->Find("price"));
+  ASSERT_TRUE(parsed->Find("simpleint"));
+  ASSERT_TRUE(parsed->Find("maxint"));
+  ASSERT_TRUE(parsed->Find("maxuint"));
+  ASSERT_TRUE(parsed->Find("minint"));
+  ASSERT_TRUE(parsed->Find("minuint"));
+  ASSERT_TRUE(parsed->Find("maxint")->is_blob());
+  ASSERT_TRUE(parsed->Find("maxuint")->is_blob());
+  ASSERT_TRUE(parsed->Find("minint")->is_blob());
+  ASSERT_TRUE(parsed->Find("minuint")->is_int())
+      << "This value should stay as an int, as it fits into it";
+
+  EXPECT_EQ(*parsed->FindBool("boolean"), true);
+  EXPECT_EQ(*parsed->FindString("name"), "whatever");
+  EXPECT_EQ(*parsed->FindDouble("price"), 1.5);
+  EXPECT_EQ(*parsed->FindInt("simpleint"), std::numeric_limits<int>::max());
+  EXPECT_EQ(base::as_string_view(*parsed->FindBlob("maxint")),
+            base::NumberToString(std::numeric_limits<int64_t>::max()));
+  EXPECT_EQ(base::as_string_view(*parsed->FindBlob("maxuint")),
+            base::NumberToString(std::numeric_limits<uint64_t>::max()));
+  EXPECT_EQ(base::as_string_view(*parsed->FindBlob("minint")),
+            base::NumberToString(std::numeric_limits<int64_t>::min()));
+  EXPECT_EQ(*parsed->FindInt("minuint"), 0);
+}
```

Comment:
```
TEST(Json64BitSupportTest, Read64BitValues2) {
  EXPECT_EQ(*base::JSONReader::ReadDict(kSampleJson)->FindDouble("maxuint"),
            std::numeric_limits<uint64_t>::max());

  EXPECT_EQ(
      *base::JSONReader::ReadDict(kSampleJson, base::JSON_ALLOW_64BIT_NUMBERS)
           ->FindDouble("maxuint"),
      std::numeric_limits<uint64_t>::max());
}
```
I'm afraid such test fails meaning that behavior of extracting doubles from base::Value is getting changed. Afaik for some remote API payloads wallet has to support both uint64 and double.

Reply:
The you wrote is wrong:

```
  EXPECT_EQ(*base::JSONReader::ReadDict(kSampleJson)->FindDouble("maxuint"),
            std::numeric_limits<uint64_t>::max());
```

This needs to parse with `base::JSON_ALLOW_64BIT_NUMBERS`

```
  EXPECT_EQ(
      *base::JSONReader::ReadDict(kSampleJson, base::JSON_ALLOW_64BIT_NUMBERS)
           ->FindDouble("maxuint"),
      std::numeric_limits<uint64_t>::max());
```

`maxuint` is a uint64_t value and won't be read as a double.

Reply:
To clarify a bit further. Double support has not been removed, nor does it conflict here. The only difference now is that values that are 64 bit and are not marked as double, are treated as int64.

Reply:
In fact we don't need that many cases to support uint64 fields when parsing JSON RPC responses. Many of them come from js world where they are integer `number`s and are limited to `Number.MAX_SAFE_INTEGER`. So having `std::optional<uint64_t> base::Value::GetSafeInteger()` backed by double would be enough for many cases.

Reply:
Not sure why the test is wrong. That's a bit simpler which should not fail:
```
TEST(Json64BitSupportTest, Read64BitValues2) {
  EXPECT_TRUE(base::JSONReader::ReadDict(kSampleJson)->FindDouble("maxuint"));
  EXPECT_TRUE(
      base::JSONReader::ReadDict(kSampleJson, base::JSON_ALLOW_64BIT_NUMBERS)
          ->FindDouble("maxuint"));
}
```
Having `base::JSON_ALLOW_64BIT_NUMBERS` means *every* `double` value might become not a double but a blob and that's seems a breaking change for large payloads where actual large doubles are mixed with possible uint64 integers

Reply:
> Having base::JSON_ALLOW_64BIT_NUMBERS means every double value might become not a double but a blob

No that's now how it works. This is parsing just fine as a double.

```cxx
EXPECT_EQ(*parsed->FindDouble("price"), 1.5);
```
The only thing that is changing is interger promotion from int to double, but a double value in json has a `.` notation and that stays unaffected.

I think you misunderstood the problem. After this PR we are not gonna use `base::Value` directly either. I'm plan to go over wherever we are using `base::Value` and replace it with a type generated with the json schema compiler.

Reply:
I don't think we should ever assume that a double can correctly hold a 64 bit int as a general case because it can only hold every integer up to 2^53. We may be able to use that in some specific cases if we are forced into it for some reason, but generally speaking I think using doubles to represent 64 bit integers of any value is a really bad idea

Reply:
My plan is to have the json schema compiler scooping over these values. I can add some additional handling in the json schema compiler for int64 values that may fit on a double, if the field in the schema is tagged as double, but that's all implementation detail.